src/builder.rs

```diff
@@ -6,18 +6,32 @@ use utils::Toggle;
 use params::NoiseParams;
 use resolvers::CryptoResolver;
 use error::{SnowError, InitStage, Prerequisite};
+use subtle::ConstantTimeEq;
 
 /// A keypair object returned by [`generate_keypair()`]
 ///
 /// [`generate_keypair()`]: #method.generate_keypair
-#[derive(PartialEq)]
 pub struct Keypair {
     /// The private asymmetric key
     pub private: Vec<u8>,
     /// The public asymmetric key
     pub public: Vec<u8>,
 }
 
+
+impl PartialEq for Keypair {
+    fn eq(&self, other: &Keypair) -> bool {
+        let priv_eq = self.private.ct_eq(&other.private).unwrap_u8() == 1u8;
+        let pub_eq = self.public.ct_eq(&other.public).unwrap_u8() == 1u8;
+
+        if priv_eq && pub_eq {
```

Comment:
Honestly I still don't fully understand how to beat rust/LLVM's optimizer for ensuring constant-timeness, but should this instead be something like

```rust
if priv_eq & pub_eq {
```

to avoid short-circuit evaluation?

Reply:
I'm not sure if it can be beaten at all. Nice catch! Even though using `&&` shouldn't leak any details about the private key, it could be used to determine which of the keys was incorrect.